### DISCUSSION TITLE
[Python] Add way to select columns by type

### DISCUSSION THREAD

**Discussion by Tishj** [duckdb/duckdb#6706] [Ideas]
Recently answered on Discord, would you be interested in being able to project columns from a relation based on the type of the column?
```py
import duckdb

con = duckdb.sql("create table tbl(firstname VARCHAR, lastname VARCHAR, age BIGINT, date_of_birth TIMESTAMP, place_of_birth VARCHAR)");
# Only project the columns of the table with this LogicalType
rel = con.table("tbl").project(duckdb.VARCHAR)
# Selected only `firstname`, `lastname` and `place_of_birth` from the table
```

**Comment by alexboks** [comment]
It would be nice if this allowed you to select multiple types. (e.g. cases where you have multiple numeric types)

**Reply by Tishj** [reply]
I think that can also be supported, probably through a list of types, or maybe we can support varargs for this, but that seems unnecessary

This should also be possible:
```py
con.table("tbl").project([duckdb.VARCHAR, duckdb.BIGINT])
```

**Reply by alexboks** [reply]
How would this resolve?
`con.table("tbl").project([duckdb.VARCHAR, duckdb.BIGINT]).sql_query()`
It just figures out the columns?

**Reply by Tishj** [reply]
so relations already contain a list of columns, which already contain a type. So this would just mean filtering that list

**Reply by Tishj** [reply]
I have this working locally now:
```py
import duckdb

duckdb.sql("create table tbl(firstname VARCHAR, lastname VARCHAR, age BIGINT, date_of_birth TIMESTAMP, place_of_birth VARCHAR)");
# Only project the columns of the table with this LogicalType
rel = duckdb.table("tbl").project(["varchar"])
rel.show()
# Selected only `firstname`, `lastname` and `place_of_birth` from the table

"""
┌───────────┬──────────┬────────────────┐
│ firstname │ lastname │ place_of_birth │
│  varchar  │ varchar  │    varchar     │
├───────────────────────────────────────┤
│                0 rows                 │
└───────────────────────────────────────┘
"""

```

**Reply by Tishj** [reply]
```py
rel = duckdb.table("tbl").project(["varchar", "bigint"])
rel.show()
```
```
┌───────────┬──────────┬───────┬────────────────┐
│ firstname │ lastname │  age  │ place_of_birth │
│  varchar  │ varchar  │ int64 │    varchar     │
├───────────────────────────────────────────────┤
│                    0 rows                     │
└───────────────────────────────────────────────┘
```

**Reply by Tishj** [reply]
Most of the implementation work left would be creating an enum out of our basic types, so we can support `duckdb.VARCHAR`, and a follow up to that would likely be turning that into a class instead so we can also support constructing STRUCT or LIST types

**Reply by Alex-Monahan** [reply]
Would you be open to adding this to the `COLUMNS` syntax in SQL?

**Reply by Tishj** [reply]
Yea I think this should be a temporary implementation and indeed COLUMNS should be extended to support this 👍 
It *should* be relatively easy because of the work done in <https://github.com/duckdb/duckdb/pull/6621>

I don't think it can be implemented in a backwards compatible manner though, the newly introduced lambda syntax would have to be extended, to support:
```sql
SELECT COLUMNS(x -> x.type = "int64") FROM test;
┌───────┬───────┬───────┐
│  c1   │  c3   │  c5   │
│ int64 │ int64 │ int64 │
├───────┼───────┼───────┤
│    10 │    12 │    14 │
│    20 │    22 │    24 │
└───────┴───────┴───────┘
```
Turning the existing example in the linked PR into:
```sql
SELECT COLUMNS(x -> x.name[-1] IN (1, 3, 5)) FROM test;
┌───────┬───────┬───────┐
│  c1   │  c3   │  c5   │
│ int64 │ int64 │ int64 │
├───────┼───────┼───────┤
│    10 │    12 │    14 │
│    20 │    22 │    24 │
└───────┴───────┴───────┘
```

**Reply by Mytherin** [reply]
I think turning the lambda into a struct is a great idea. Also because currently the lambda emits qualified names as flattened varchars (e.g. `'test.c1', 'test.c2'`) which I'm not too happy about. Turning it into a struct would allow this to be `x.table, x.name, x.type`. Backwards compatibility is not required yet here because it is not yet a released feature.

The only remaining question is how do you do the result binding when you get a list like `['c1', 'VARCHAR']`. I propose that you cannot mix types and column names, and you start off by binding the first element - if it is an existing column, we assume everything is a column name. If it is a type - all elements are types. Otherwise you throw an error (column not found).

**Comment by osalloum** [comment]
This actually would be very useful when creating parquet files , a few engines do not support the UUID type which needs to be casted as varchar in that case

**Reply by Tishj** [reply]
> This actually would be very useful when creating parquet files , a few engines do not support the UUID type which needs to be casted as varchar in that case

That has been added as `select_types` on the DuckDBPyRelation
